pkg/server/ca/upstream_client.go

```diff
@@ -0,0 +1,339 @@
+	})
+	defer func() {
+		if err != nil {
+			u.publishJWTKeyStream.Stop()
+		}
+	}()
+
+	select {
+	case result := <-firstResultCh:
+		return result.jwtKeys, result.err
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	}
+}
+
+func (u *UpstreamClient) runMintX509CAStream(ctx context.Context, req *upstreamauthority.MintX509CARequest, firstResultCh chan<- mintX509CAResult) {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	stream, err := u.c.UpstreamAuthority.MintX509CA(ctx, req)
+	if err != nil {
+		firstResultCh <- mintX509CAResult{err: err}
+		return
+	}
+
+	resp, err := stream.Recv()
+	if err != nil {
+		firstResultCh <- mintX509CAResult{err: err}
+		return
+	}
+
+	x509CA, x509Roots, err := parseMintX509CAFirstResponse(resp)
+	if err != nil {
+		firstResultCh <- mintX509CAResult{err: err}
+		return
+	}
+
+	if !u.c.UpstreamBundle {
+		// We have opted not to join the upstream PKI. The server CA should
+		// therefore be considered the root. Also, this means there is no
+		// reason to continue monitoring for bundle updates, so update the
+		// bundle with the server CA and return.
+		if err := u.c.BundleUpdater.AppendX509Roots(ctx, x509CA[:1]); err != nil {
+			firstResultCh <- mintX509CAResult{err: err}
+			return
+		}
+		firstResultCh <- mintX509CAResult{x509CA: x509CA, done: true}
+		return
+	}
+
+	if err := u.c.BundleUpdater.AppendX509Roots(ctx, x509Roots); err != nil {
+		firstResultCh <- mintX509CAResult{err: err}
+		return
+	}
+
+	firstResultCh <- mintX509CAResult{x509CA: x509CA}
+
+	for {
+		resp, err := stream.Recv()
+		if err != nil {
+			if err != io.EOF {
+				u.c.BundleUpdater.OnError(err, "MintX509CA stream closed prematurely")
+			}
+			return
+		}
+
+		x509Roots, err := parseMintX509CABundleUpdate(resp)
+		if err != nil {
+			u.c.BundleUpdater.OnError(err, "Invalid MintX509CA bundle update response")
+			continue
+		}
+
+		if err := u.c.BundleUpdater.AppendX509Roots(ctx, x509Roots); err != nil {
+			u.c.BundleUpdater.OnError(err, "Failed to append upstream X.509 roots")
+			continue
+		}
+	}
+}
+
+func (u *UpstreamClient) runPublishJWTKeyStream(ctx context.Context, req *upstreamauthority.PublishJWTKeyRequest, firstResultCh chan<- publishJWTKeyResult) {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	stream, err := u.c.UpstreamAuthority.PublishJWTKey(ctx, req)
+	if err != nil {
+		firstResultCh <- publishJWTKeyResult{err: err}
+		return
+	}
+
+	resp, err := stream.Recv()
+	if err != nil {
+		firstResultCh <- publishJWTKeyResult{err: err}
+		return
+	}
+
+	updatedKeys, err := u.c.BundleUpdater.AppendJWTKeys(ctx, resp.UpstreamJwtKeys)
+	if err != nil {
+		firstResultCh <- publishJWTKeyResult{err: err}
+		return
+	}
+
+	firstResultCh <- publishJWTKeyResult{jwtKeys: updatedKeys}
+
+	for {
+		resp, err := stream.Recv()
+		if err != nil {
+			if err != io.EOF {
+				u.c.BundleUpdater.OnError(err, "PublishJWTKey stream closed prematurely")
+			}
+			return
```

Comment:
can you add a comment about why it is not creating logs for this err? (same in mint method)

Reply:
I think instead I'll change the interface method from `OnError` to `LogError`.... that makes it clear what the implementation will be doing.

Reply:
I think @MarcosDY was asking for the reason that io.EOF closes the stream without logging anything

Reply:
yeah I was asking about that thanks for point that Evan

Reply:
Oh right :) I'll add a comment.